### PR TITLE
Fix mypy for `utilities.memory`

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

Part of #7037. This PR disentangles large #7218. It fixes mypy typing in `utilities.memory.py`.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@stancld thanks for doing these! very appreciated <3
how much of this file is WIP?

**Comment by stancld** [comment]
@awaelchli No worries, it's great to learn something new :) There's one last mypy issue.
```
pytorch_lightning/utilities/memory.py:139: error: List item 0 has incompatible type "Optional[str]"; expected "Union[str, bytes, PathLike[str], PathLike[bytes]]"  [list-item]
```

**Review by carmocca — APPROVED** [review]
> You just need to check that `shutil.which` does not return `None`

File: pytorch_lightning/utilities/memory.py [hunk 1]
```diff
@@ -135,8 +135,11 @@ def get_gpu_memory_map() -> Dict[str, float]:
         A dictionary in which the keys are device ids as integers and
         values are memory usage as integers in MB.
     """
+    nvidia_smi_path = shutil.which("nvidia-smi")
+    if nvidia_smi_path is None:
+        raise FileNotFoundError("nvidia-smi: command not found")
```
Comment (ananthsub) [inline pytorch_lightning/utilities/memory.py:140]: could you also update the docstring for the function that this raises a FileNotFoundError?

**Review by awaelchli — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Pull request merged**